Use `setup.py` to build and install vs. bash scripts

gnestor opened the pull request:
Closes https://github.com/jupyterlab/mimerender-cookiecutter/issues/38

gnestor committed:
Use ActivityMonitor to throttle users typing to update model

gnestor committed:
Merge pull request #40 from gnestor/activity-monitor

Use ActivityMonitor to throttle users typing to update model

gnestor committed:
Patch: Import ActivityMonitor

gnestor committed:
Merge pull request #42 from gnestor/activity-monitor

Patch: Import ActivityMonitor

gnestor committed:
Use RequireJS vs. jQuery

gnestor commented on {{cookiecutter.extension_name}}/setup.py:
```diff
@@ -1,38 +1,197 @@
 
+repo_root = os.path.dirname(os.path.abspath(__file__))
+sym_link = '--symlink' if sys.platform != 'win32' else ''
+sys_prefix = '--sys-prefix' if os.environ.get('CONDA_DEFAULT_ENV', os.defpath) else ''
```
Is this an effective way to determine whether to use the `--sys-prefix` flag?

gnestor commented on {{cookiecutter.extension_name}}/setup.py:
```diff
@@ -1,38 +1,197 @@
+    build_nbextension()
     setup(**setup_args)
+    install_labextension()
+    install_nbextension()
```
Should install and enable extensions commands be run here?

gnestor commented:
@ellisonbg Here is my first go at moving all of the build/install scripts into `setup.py`. I tested and it seems to be working, however I've never written a `setup.py`, so please advise about how it could be improved or feel free to invite others to review.

gnestor committed:
Use sub-class of IPython.display.JSON vs. display function

gnestor committed:
Add workflow section to README

gnestor committed:
Merge pull request #48 from gnestor/readme

Add workflow section to README

gnestor committed:
Merge pull request #44 from gnestor/requirejs

Use RequireJS vs. jQuery

gnestor committed:
0.16.0

jupyterlab@^0.16.0

gnestor committed:
Merge pull request #51 from gnestor/master

0.16.0

gnestor committed:
Use `setup.py` to build and install vs. bash scripts

gnestor committed:
Normalize paths in webpack config

gnestor committed:
Move tasks to sub-classes of `develop` and `install` commands

gnestor committed:
Merge pull request #47 from gnestor/display-class

Use sub-class of IPython.display.JSON vs. display function

gnestor committed:
Merge pull request #46 from gnestor/webpack-windows

Normalize paths in webpack config

gnestor committed:
Patch #44: Use RequireJS vs. jQuery

gnestor committed:
Merge pull request #54 from gnestor/master

Patch #44: Use RequireJS vs. jQuery

ellisonbg commented:
So far it has been our "policy" and recommendation to not install or enable lab/nb extensions from setup.py scripts. The reason is that a user may have multiple extensions (a theme for example) that does the same thing. Because of this, we are currently recommending that packagers *force* users to do the separate lab/nb extension installation. Sorry I wasn't clear about that before.

But this means that your setup.py can be simpler. It should do `npm install` and `npm build` in the nb/lab extension, but that is it.

ellisonbg commented:
And to keep things simple, I would honestly just copy/port the relevant NPM using code from the jupyterlab build scripts:

https://github.com/jupyterlab/jupyterlab/blob/master/setup.py
https://github.com/jupyterlab/jupyterlab/blob/master/setupbase.py#L103

What would *really* help is to have a separate Python package that had the build utilities. Then all of our JS projects (at least the new ones) could use those. @blink1073 @afshin @jasongrout @SylvainCorlay how does that sound? Basically abstract out the setup.py and setupbase.py of jupyterlab into a separate project we can use throughout the lab ecosystem?

blink1073 commented:
Ha, `jupyter/apackagemanagerhasnoname`.   It would at least consolidate the ickiness.

gnestor committed:
Clean up

gnestor commented:
@ellisonbg I removed the "install and enable extension" command and cleaned up the rest using https://github.com/jupyterlab/jupyterlab/blob/master/setupbase.py as reference.

I tried porting over the `NPM` command class and adding it as `jsdeps` command, but it isn't run when doing `pip install -e .`. I created a WIP branch: https://github.com/gnestor/mimerender-cookiecutter/tree/jlab-setup. How do you tell setuptools to run a command pre/post-install (other than sub-classing the `develop` or `install` commands)?

gnestor committed:
Adapt new jupyterlab renderer interface

gnestor committed:
0.17.0

jupyterlab@^0.17

gnestor committed:
Upgrade phosphor

ellisonbg commented:
@gnestor I am definitely not an expert in setuptools/distutils. And I know its APIs are not very transparent. But I think some of the special magic is here:

https://github.com/jupyterlab/jupyterlab/blob/master/setup.py#L100

If you continue to have trouble, maybe let's setup a meeting with @blink1073 to go over the best way of capturing this logic for our different repos.

blink1073 commented:
This project looks like a good reference point for some of the subtleties of such a tool: https://github.com/ccpgames/setuphelpers/blob/master/README.rst

gnestor committed:
Catch errors when rendering document content and display error

gnestor commented:
@ellisonbg I saw that line in jupyterlab's `setup.py` but simply adding a `jsdeps` command to `setup_args.cmd_class` doesn't make it run on `pip install -e .`. I found that adding a `develop` or `install` command does though.

@blink1073 setuphelpers looks good, but between the classic notebook's and jupyterlab' `setupbase.py`, it looks like we have most of these tools already. 

As for next steps: I can meet with @blink1073 or whoever has the most experience with writing these `setup.py`s, or we can merge what I have here. The only significant difference between the `NPM` command in jupyterlab and this is [these lines](https://github.com/jupyterlab/jupyterlab/blob/master/setupbase.py#L136-L141) which verify that the required files were built...which I can add if necessary.

blink1073 commented:
I'd say we should go ahead and make the repo, and we can iterate on the content.

gnestor committed:
Include display metadata in mime bundle

gnestor committed:
Fix nbextension styles

gnestor committed:
Persist cookiecutter inputs to `.cookiecutter.yaml`

gnestor committed:
Update workflow section of README and RELEASE

gnestor committed:
Merge pull request #55 from gnestor/0.17

0.17.0

gnestor commented:
@blink1073 Sounds good to me. Would you like to get create it?

blink1073 commented:
I'm mid stream on too many things at the moment.

gnestor commented:
Ok, should I create it in jupyter or jupyterlab org?

blink1073 commented:
Seems like it jupyter since nbdime and notebook could possibly use it

gnestor commented:
@ellisonbg It looks like I don't have sufficient permissions to create a repo in the jupyter org. Can you create one for this and "grant" @blink1073 and myself push rights?

ellisonbg commented:
How shall I name it?

On Thu, Mar 2, 2017 at 10:16 AM, Grant Nestor <notifications@github.com>
wrote:

> @ellisonbg <https://github.com/ellisonbg> It looks like I don't have
> sufficient permissions to create a repo in the jupyter org. Can you create
> one for this and "grant" @blink1073 <https://github.com/blink1073> and
> myself push rights?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jupyterlab/mimerender-cookiecutter/pull/45#issuecomment-283734255>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABr0KE6qe8nO6-mnJvgHk35H9xKWjTpks5rhweEgaJpZM4MCmY_>
> .
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

gnestor commented:
I'm not sure... Something to do with "setup"/"setuptools" and "jupyter" and "js". 

I found some good examples:

* https://github.com/jupyter-widgets/widget-cookiecutter/blob/master/%7B%7Bcookiecutter.github_project_name%7D%7D/setup.py
* https://github.com/ipython/ipywidgets/blob/master/widgetsnbextension/setup.py
* https://github.com/ipython/ipywidgets/blob/master/jupyterlab_widgets/setup.py

ellisonbg commented:
Maybe jupyter-js-setup?

On Sat, Mar 4, 2017 at 8:28 AM, Grant Nestor <notifications@github.com>
wrote:

> I'm not sure... Something to do with "setup"/"setuptools" and "jupyter"
> and "js".
>
> I found some good examples:
>
>    - https://github.com/jupyter-widgets/widget-cookiecutter/
>    blob/master/%7B%7Bcookiecutter.github_project_name%7D%7D/setup.py
>    <https://github.com/jupyter-widgets/widget-cookiecutter/blob/master/%7B%7Bcookiecutter.github_project_name%7D%7D/setup.py>
>    - https://github.com/ipython/ipywidgets/blob/master/
>    widgetsnbextension/setup.py
>    - https://github.com/ipython/ipywidgets/blob/master/
>    jupyterlab_widgets/setup.py
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jupyterlab/mimerender-cookiecutter/pull/45#issuecomment-284162404>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABr0EyjAv1a0tQvlIH--D2p41Dh8zXjks5riZFFgaJpZM4MCmY_>
> .
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

blink1073 commented:
Why not jupyter-distutils and make it generic for both python only and those that use npm?





On Sat, Mar 4, 2017 at 11:56 AM -0600, "Brian E. Granger" <notifications@github.com> wrote:










Maybe jupyter-js-setup?



On Sat, Mar 4, 2017 at 8:28 AM, Grant Nestor <notifications@github.com>

wrote:



> I'm not sure... Something to do with "setup"/"setuptools" and "jupyter"

> and "js".

>

> I found some good examples:

>

>    - https://github.com/jupyter-widgets/widget-cookiecutter/

>    blob/master/%7B%7Bcookiecutter.github_project_name%7D%7D/setup.py

>    <https://github.com/jupyter-widgets/widget-cookiecutter/blob/master/%7B%7Bcookiecutter.github_project_name%7D%7D/setup.py>

>    - https://github.com/ipython/ipywidgets/blob/master/

>    widgetsnbextension/setup.py

>    - https://github.com/ipython/ipywidgets/blob/master/

>    jupyterlab_widgets/setup.py

>

> —

> You are receiving this because you were mentioned.

> Reply to this email directly, view it on GitHub

> <https://github.com/jupyterlab/mimerender-cookiecutter/pull/45#issuecomment-284162404>,

> or mute the thread

> <https://github.com/notifications/unsubscribe-auth/AABr0EyjAv1a0tQvlIH--D2p41Dh8zXjks5riZFFgaJpZM4MCmY_>

> .

>







-- 

Brian E. Granger

Associate Professor of Physics and Data Science

Cal Poly State University, San Luis Obispo

@ellisonbg on Twitter and GitHub

bgranger@calpoly.edu and ellisonbg@gmail.com





—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

ellisonbg commented:
Great, point, I will create that now!

On Sat, Mar 4, 2017 at 11:35 AM, Steven Silvester <notifications@github.com>
wrote:

> Why not jupyter-distutils and make it generic for both python only and
> those that use npm?
>
>
>
>
>
> On Sat, Mar 4, 2017 at 11:56 AM -0600, "Brian E. Granger" <
> notifications@github.com> wrote:
>
>
>
>
>
>
>
>
>
>
> Maybe jupyter-js-setup?
>
>
>
> On Sat, Mar 4, 2017 at 8:28 AM, Grant Nestor <notifications@github.com>
>
> wrote:
>
>
>
> > I'm not sure... Something to do with "setup"/"setuptools" and "jupyter"
>
> > and "js".
>
> >
>
> > I found some good examples:
>
> >
>
> > - https://github.com/jupyter-widgets/widget-cookiecutter/
>
> > blob/master/%7B%7Bcookiecutter.github_project_name%7D%7D/setup.py
>
> > <https://github.com/jupyter-widgets/widget-cookiecutter/blob/master/%7B%
> 7Bcookiecutter.github_project_name%7D%7D/setup.py>
>
> > - https://github.com/ipython/ipywidgets/blob/master/
>
> > widgetsnbextension/setup.py
>
> > - https://github.com/ipython/ipywidgets/blob/master/
>
> > jupyterlab_widgets/setup.py
>
> >
>
> > —
>
> > You are receiving this because you were mentioned.
>
> > Reply to this email directly, view it on GitHub
>
> > <https://github.com/jupyterlab/mimerender-cookiecutter/pull/45#
> issuecomment-284162404>,
>
> > or mute the thread
>
> > <https://github.com/notifications/unsubscribe-auth/AABr0EyjAv1a0tQvlIH--
> D2p41Dh8zXjks5riZFFgaJpZM4MCmY_>
>
> > .
>
> >
>
>
>
>
>
>
>
> --
>
> Brian E. Granger
>
> Associate Professor of Physics and Data Science
>
> Cal Poly State University, San Luis Obispo
>
> @ellisonbg on Twitter and GitHub
>
> bgranger@calpoly.edu and ellisonbg@gmail.com
>
>
>
>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>
>
>
>
>
>
>
>
>
>
>
>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jupyterlab/mimerender-cookiecutter/pull/45#issuecomment-284175426>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABr0D23ltQULzdWc0L_Rpy2XswNx94mks5rib0AgaJpZM4MCmY_>
> .
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

ellisonbg commented:
https://github.com/jupyter/jupyter-distutils

I have invited grant and steve as repo admins...

On Sat, Mar 4, 2017 at 11:36 AM, Brian Granger <ellisonbg@gmail.com> wrote:

> Great, point, I will create that now!
>
> On Sat, Mar 4, 2017 at 11:35 AM, Steven Silvester <
> notifications@github.com> wrote:
>
>> Why not jupyter-distutils and make it generic for both python only and
>> those that use npm?
>>
>>
>>
>>
>>
>> On Sat, Mar 4, 2017 at 11:56 AM -0600, "Brian E. Granger" <
>> notifications@github.com> wrote:
>>
>>
>>
>>
>>
>>
>>
>>
>>
>>
>> Maybe jupyter-js-setup?
>>
>>
>>
>> On Sat, Mar 4, 2017 at 8:28 AM, Grant Nestor <notifications@github.com>
>>
>> wrote:
>>
>>
>>
>> > I'm not sure... Something to do with "setup"/"setuptools" and "jupyter"
>>
>> > and "js".
>>
>> >
>>
>> > I found some good examples:
>>
>> >
>>
>> > - https://github.com/jupyter-widgets/widget-cookiecutter/
>>
>> > blob/master/%7B%7Bcookiecutter.github_project_name%7D%7D/setup.py
>>
>> > <https://github.com/jupyter-widgets/widget-cookiecutter/blob
>> /master/%7B%7Bcookiecutter.github_project_name%7D%7D/setup.py>
>>
>> > - https://github.com/ipython/ipywidgets/blob/master/
>>
>> > widgetsnbextension/setup.py
>>
>> > - https://github.com/ipython/ipywidgets/blob/master/
>>
>> > jupyterlab_widgets/setup.py
>>
>> >
>>
>> > —
>>
>> > You are receiving this because you were mentioned.
>>
>> > Reply to this email directly, view it on GitHub
>>
>> > <https://github.com/jupyterlab/mimerender-cookiecutter/pull/
>> 45#issuecomment-284162404>,
>>
>> > or mute the thread
>>
>> > <https://github.com/notifications/unsubscribe-auth/
>> AABr0EyjAv1a0tQvlIH--D2p41Dh8zXjks5riZFFgaJpZM4MCmY_>
>>
>> > .
>>
>> >
>>
>>
>>
>>
>>
>>
>>
>> --
>>
>> Brian E. Granger
>>
>> Associate Professor of Physics and Data Science
>>
>> Cal Poly State University, San Luis Obispo
>>
>> @ellisonbg on Twitter and GitHub
>>
>> bgranger@calpoly.edu and ellisonbg@gmail.com
>>
>>
>>
>>
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub, or mute the thread.
>>
>>
>>
>>
>>
>>
>>
>>
>>
>>
>>
>>
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/jupyterlab/mimerender-cookiecutter/pull/45#issuecomment-284175426>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AABr0D23ltQULzdWc0L_Rpy2XswNx94mks5rib0AgaJpZM4MCmY_>
>> .
>>
>
>
>
> --
> Brian E. Granger
> Associate Professor of Physics and Data Science
> Cal Poly State University, San Luis Obispo
> @ellisonbg on Twitter and GitHub
> bgranger@calpoly.edu and ellisonbg@gmail.com
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

gnestor commented:
Thanks!

gnestor committed:
Use jupyter-packaging

gnestor commented:
@blink1073 How would I add jupyter-packaging to the `devDependencies` for this repo? From my understanding, `install_requires` specifies the _production_ dependencies but what about the dependencies for development and distribution?

blink1073 commented:
I'll ping you after I do it in the jlab PR. Headed to bed.

blink1073 commented:
Closing in favor of #64.
